metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java

```diff
@@ -355,44 +356,58 @@ private void loadEntities(EntityType entityType, MongoTerm term, Jedis jedis,
   /**
    * Get an enrichment document based on the uri provided.
    *
-   * @param uri the provided uri
+   * @param requestedUri the provided uri
    * @return the enrichment document corresponding to the provided uri
    * @throws IOException if something went wrong when accessing the db
    */
-  public EntityWrapper getByUri(String uri) throws IOException {
-    Jedis jedis = redisProvider.getJedis();
-    EntityWrapper entityWrapper = null;
-    entityWrapper = getEntityWrapper(uri, jedis, entityWrapper, CACHED_AGENT);
-    entityWrapper = getEntityWrapper(uri, jedis, entityWrapper, CACHED_CONCEPT);
-    entityWrapper = getEntityWrapper(uri, jedis, entityWrapper, CACHED_TIMESPAN);
-    entityWrapper = getEntityWrapper(uri, jedis, entityWrapper, CACHED_PLACE);
-
-    entityWrapper = getEntityWrapperFromSameAs(uri, jedis, entityWrapper, CACHED_AGENT);
-    entityWrapper = getEntityWrapperFromSameAs(uri, jedis, entityWrapper, CACHED_CONCEPT);
-    entityWrapper = getEntityWrapperFromSameAs(uri, jedis, entityWrapper, CACHED_TIMESPAN);
-    entityWrapper = getEntityWrapperFromSameAs(uri, jedis, entityWrapper, CACHED_PLACE);
-    jedis.close();
-    return entityWrapper;
+  public EntityWrapper getByUri(String requestedUri) throws IOException {
+
+    // Get the result providers - this is constant and does not depend on the input.
+    final List<GetterByUri> getters = Arrays.asList(
+            (uri, jedis) -> getEntityWrapper(uri, jedis, CACHED_AGENT),
+            (uri, jedis) -> getEntityWrapper(uri, jedis, CACHED_CONCEPT),
+            (uri, jedis) -> getEntityWrapper(uri, jedis, CACHED_TIMESPAN),
+            (uri, jedis) -> getEntityWrapper(uri, jedis, CACHED_PLACE),
+            (uri, jedis) -> getEntityWrapperFromSameAs(uri, jedis, CACHED_AGENT),
+            (uri, jedis) -> getEntityWrapperFromSameAs(uri, jedis, CACHED_CONCEPT),
+            (uri, jedis) -> getEntityWrapperFromSameAs(uri, jedis, CACHED_TIMESPAN),
+            (uri, jedis) -> getEntityWrapperFromSameAs(uri, jedis, CACHED_PLACE)
+    );
+
+    // Get the first non-null result.
+    try (final Jedis jedis = redisProvider.getJedis()) {
+      for (GetterByUri getter : getters) {
+        final EntityWrapper result = getter.getByUri(requestedUri, jedis);
+        if (result != null) {
+          return result;
```

Comment:
This actually changes behavior of the retrieved result.
Before we were getting the last found and now we are getting the first found.
Are we aware of and fine with this change?

Reply:
I was aware of it. At that point, you expect only one of these to yield any result at all anyway, so it doesn't matter. And even if it does, we would like to return the record with the matching rdf:about value, instead of the record with the matching sameAs value.